# ezRAD-Pool-seq

## An implementation of the ezRAD pipeline described in Phair _et. al._ (2019).

#### Downloading the _Z. marina_ genome from NCBI Genomes. 

The [_Zostera marina_](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=29655&lvl=3&lin=f&keep=1&srchmode=1&unlock) assembly is available by [ftp](https://ftp.ncbi.nih.gov/genomes/genbank/plant/Zostera_marina/all_assembly_versions/GCA_001185155.1_Zosma_marina.v.2.1/). The [genomic.fna.gz](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/185/155/GCA_001185155.1_Zosma_marina.v.2.1/GCA_001185155.1_Zosma_marina.v.2.1_genomic.fna.gz). Interesting piece on contigs and scaffolds [here](https://www.pacb.com/blog/genomes-vs-gennnnes-difference-contigs-scaffolds-genome-assemblies/).


#### Indexing the reference genome.

Create an index of the reference genome using the the BWA algorithm (https://github.com/lh3/bwa)

```
  bwa index GCA_001185155.1_Zosma_marina.v.2.1_genomic.fna

```
This index is used in mapping of reads to reference step.


#### Mapping reads to the reference 

The gzipped FastQ reads are mapped to the reference genome generated by the BWA algorithm to generate SAM files with minimum score of 20 parameter. The location of the BWA index is set in the _params.bwa_ref_ parameter of the nextflow file. SAM files are converted to BAM files using [SAMtools](http://www.htslib.org/doc/samtools.html) [samtools cheatsheet](https://coding4medicine.com/cheatsheets/samtools.html).
Mapped reads and unmapped reads were calculated using the samtools _idxstats_ option which reports the summary statistics






## Citations ##

Phair NL, Toonen RJ, Knapp I, von der Heyden S. 2019. Shared genomic outliers across two divergent population clusters of a highly threatened seagrass. PeerJ 7:e6806 http://doi.org/10.7717/peerj.6806
